File: kelmah-mobile-ios/Kelmah/Features/Profile/Data/ProfileRepository.swift
import Foundation

final class ProfileRepository {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func getWorkerProfileSnapshot(workerId: String) async throws -> WorkerProfileSnapshot {
        do {
            let response = try await apiClient.send(
                path: "users/me/profile-signals",
                method: .get,
                requiresAuth: true,
                responseType: ProfileRawEnvelope.self
            )
            return parseProfileSignals(response)
        } catch let APIClientError.invalidStatusCode(statusCode, _) where statusCode == 404 {
            return try await getLegacyWorkerProfileSnapshot(workerId: workerId)
        }
    }

    private func getLegacyWorkerProfileSnapshot(workerId: String) async throws -> WorkerProfileSnapshot {
        async let credentialsResponse = optionalFetch(path: "users/me/credentials")
        async let availabilityResponse = optionalFetch(path: "users/workers/\(workerId.urlPathEncoded)/availability")
        async let completenessResponse = optionalFetch(path: "users/workers/\(workerId.urlPathEncoded)/completeness")
        async let portfolioResponse = optionalFetch(
            path: "users/workers/\(workerId.urlPathEncoded)/portfolio",
            queryItems: [URLQueryItem(name: "limit", value: "6")]
        )

        let profileResponse = try await apiClient.send(
            path: "users/profile",
            method: .get,
            requiresAuth: true,
            responseType: ProfileRawEnvelope.self
        )

        let resolvedCredentials = await credentialsResponse
        let resolvedAvailability = await availabilityResponse
        let resolvedCompleteness = await completenessResponse
        let resolvedPortfolio = await portfolioResponse

        return WorkerProfileSnapshot(
            profile: parseProfile(profileResponse),
            credentials: parseCredentials(resolvedCredentials),
            availability: parseAvailability(resolvedAvailability),
            completeness: parseCompleteness(resolvedCompleteness),
            portfolio: parsePortfolio(resolvedPortfolio),
            partialWarnings: [
                resolvedCredentials == nil ? "Your certificates could not load. Job matches may be less accurate." : nil,
                resolvedAvailability == nil ? "Your work time could not load. Job matches may be less up to date." : nil,
                resolvedCompleteness == nil ? "Your profile check could not load." : nil,
                resolvedPortfolio == nil ? "Your past work could not load. Hirers may see less proof." : nil,
            ].compactMap { $0 }
        )
    }

    private func parseProfileSignals(_ response: ProfileRawEnvelope) -> WorkerProfileSnapshot {
        let object = response.data?.objectValue ?? [:]
        return WorkerProfileSnapshot(
            profile: parseProfileObject(object["profile"]?.objectValue ?? [:]),
            credentials: parseCredentialsObject(object["credentials"]?.objectValue ?? [:]),
            availability: parseAvailabilityObject(object["availability"]?.objectValue ?? [:]),
            completeness: parseCompletenessObject(object["completeness"]?.objectValue ?? [:]),
            portfolio: parsePortfolioObject(object["portfolio"]?.objectValue ?? [:]),
            partialWarnings: []
        )
    }

    private func optionalFetch(path: String, queryItems: [URLQueryItem] = []) async -> ProfileRawEnvelope? {
        try? await apiClient.send(
            path: path,
            method: .get,
            queryItems: queryItems,
            requiresAuth: true,
            responseType: ProfileRawEnvelope.self
        )
    }

    private func parseProfile(_ response: ProfileRawEnvelope) -> WorkerRecommendationProfile {
        parseProfileObject(response.data?.objectValue ?? [:])
    }

    private func parseProfileObject(_ object: [String: JSONValue]) -> WorkerRecommendationProfile {
        let fallbackLocation = [object.string("city"), object.string("state"), object.string("country")]
            .compactMap { $0?.nilIfEmpty }
            .joined(separator: ", ")
        return WorkerRecommendationProfile(
            id: object.string("id") ?? object.string("_id"),
            firstName: object.string("firstName") ?? "",
            lastName: object.string("lastName") ?? "",
            email: object.string("email") ?? "",
            phone: object.string("phone") ?? "",
            bio: object.string("bio") ?? "",
            location: object.string("location") ?? fallbackLocation,
            profession: object.string("profession") ?? "",
            hourlyRate: object.double("hourlyRate"),
            currency: object.string("currency") ?? "GHS",
            experienceLevel: object.string("experienceLevel"),
            yearsOfExperience: object.int("yearsOfExperience"),
            skills: object["skills"]?.arrayValue?.compactMap { skill in
                skill.stringValue ?? skill.objectValue?.string("name") ?? skill.objectValue?.string("label")
            } ?? [],
            isEmailVerified: object.bool("isEmailVerified") ?? false,
            isPhoneVerified: object.bool("isPhoneVerified") ?? false
        )
    }

    private func parseCredentials(_ response: ProfileRawEnvelope?) -> WorkerCredentials {
        parseCredentialsObject(response?.data?.objectValue ?? [:])
    }

    private func parseCredentialsObject(_ object: [String: JSONValue]) -> WorkerCredentials {
        let skills = object["skills"]?.arrayValue?.enumerated().compactMap { index, value in
            let data = value.objectValue
            let name = data?.string("name") ?? data?.string("label") ?? value.stringValue
            guard let name else { return nil }
            return CredentialSkill(
                id: data?.string("id") ?? data?.string("_id") ?? "skill-\(index)-\(name)",
                name: name,
                category: data?.string("category") ?? "general",
                proficiencyLevel: data?.string("proficiencyLevel") ?? data?.string("level") ?? "intermediate",
                yearsOfExperience: data?.int("yearsOfExperience") ?? 0,
                isVerified: data?.bool("isVerified") ?? false
            )
        } ?? []

        return WorkerCredentials(
            skills: skills,
            licenses: parseCredentialItems(object["licenses"]?.arrayValue),
            certifications: parseCredentialItems(object["certifications"]?.arrayValue)
        )
    }

    private func parseCredentialItems(_ values: [JSONValue]?) -> [CredentialItem] {
        values?.enumerated().compactMap { index, value in
            guard let object = value.objectValue else { return nil }
            guard let name = object.string("name") else { return nil }
            return CredentialItem(
                id: object.string("id") ?? object.string("_id") ?? "credential-\(index)-\(name)",
                name: name,
                issuingOrganization: object.string("issuingOrganization") ?? object.string("issuer") ?? "",
                issueDate: object.string("issueDate") ?? object.string("issuedAt"),
                expiryDate: object.string("expiryDate") ?? object.string("expiresAt"),
                status: object.string("status") ?? (object.bool("isVerified") == true ? "verified" : "pending"),
                isVerified: object.bool("isVerified") ?? false
            )
        } ?? []
    }

    private func parseAvailability(_ response: ProfileRawEnvelope?) -> WorkerAvailability {
        parseAvailabilityObject(response?.data?.objectValue ?? [:])
    }

    private func parseAvailabilityObject(_ object: [String: JSONValue]) -> WorkerAvailability {
        let schedule = (object["schedule"]?.arrayValue ?? object["daySlots"]?.arrayValue ?? []).compactMap { value -> AvailabilityDay? in
            guard let item = value.objectValue else { return nil }
            guard let day = item.string("day") else { return nil }
            let slots = (item["slots"]?.arrayValue ?? []).compactMap { slotValue -> AvailabilitySlot? in
                guard let slot = slotValue.objectValue else { return nil }
                guard let start = slot.string("start"), let end = slot.string("end") else { return nil }
                return AvailabilitySlot(start: start, end: end)
            }
            return AvailabilityDay(
                day: day,
                available: item.bool("available") ?? (slots.isEmpty == false),
                slots: slots
            )
        }

        return WorkerAvailability(
            status: object.string("status") ?? "not_set",
            isAvailable: object.bool("isAvailable") ?? (object.string("status") == "available"),
            timezone: object.string("timezone") ?? "Africa/Accra",
            schedule: schedule,
            nextAvailable: object.string("nextAvailable"),
            lastUpdated: object.string("lastUpdated"),
            message: object.string("message")
        )
    }

    private func parseCompleteness(_ response: ProfileRawEnvelope?) -> WorkerCompleteness {
        parseCompletenessObject(response?.data?.objectValue ?? [:])
    }

    private func parseCompletenessObject(_ object: [String: JSONValue]) -> WorkerCompleteness {
        return WorkerCompleteness(
            completionPercentage: object.int("completionPercentage") ?? object.int("percentage") ?? 0,
            requiredCompletion: object.int("requiredCompletion") ?? 0,
            optionalCompletion: object.int("optionalCompletion") ?? 0,
            missingRequired: object["missingRequired"]?.arrayValue?.compactMap { $0.stringValue } ?? [],
            missingOptional: object["missingOptional"]?.arrayValue?.compactMap { $0.stringValue } ?? [],
            recommendations: object["recommendations"]?.arrayValue?.compactMap { $0.stringValue } ?? []
        )
    }

    private func parsePortfolio(_ response: ProfileRawEnvelope?) -> WorkerPortfolio {
        parsePortfolioObject(response?.data?.objectValue ?? [:])
    }

    private func parsePortfolioObject(_ object: [String: JSONValue]) -> WorkerPortfolio {
        let items = object["portfolioItems"]?.arrayValue?.compactMap { value -> PortfolioProject? in
            guard let item = value.objectValue else { return nil }
            guard let id = item.string("id") ?? item.string("_id") else { return nil }
            return PortfolioProject(
                id: id,
                title: item.string("title") ?? "Untitled project",
                description: item.string("description") ?? "",
                projectType: item.string("projectType") ?? "professional",
                skillsUsed: item["skillsUsed"]?.arrayValue?.compactMap { $0.stringValue ?? $0.objectValue?.string("name") } ?? [],
                location: item.string("location"),
                clientRating: item.double("clientRating"),
                status: item.string("status") ?? "draft",
                isFeatured: item.bool("isFeatured") ?? false,
                createdAt: item.string("createdAt")
            )
        } ?? []
        let stats = object["stats"]?.objectValue
        return WorkerPortfolio(
            items: items,
            totalCount: stats?.int("total") ?? items.count,
            publishedCount: stats?.int("published") ?? items.filter { $0.status.lowercased() == "published" }.count
        )
    }
}

private extension Dictionary where Key == String, Value == JSONValue {
    func string(_ key: String) -> String? { self[key]?.stringValue }
    func int(_ key: String) -> Int? { self[key]?.intValue }
    func double(_ key: String) -> Double? { self[key]?.doubleValue }
    func bool(_ key: String) -> Bool? { self[key]?.boolValue }
}

File: kelmah-mobile-android/app/src/main/java/com/kelmah/mobile/features/profile/data/ProfileRepository.kt
package com.kelmah.mobile.features.profile.data

import com.kelmah.mobile.core.network.ApiResult
import com.kelmah.mobile.core.network.executeAuthorizedApiCall
import com.kelmah.mobile.core.session.SessionCoordinator
import dagger.Lazy
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import retrofit2.HttpException

@Singleton
class ProfileRepository @Inject constructor(
    private val profileApiService: ProfileApiService,
    private val sessionCoordinator: Lazy<SessionCoordinator>,
) {
    suspend fun getWorkerProfileSnapshot(workerId: String): ApiResult<WorkerProfileSnapshot> = executeAuthorizedApiCall(sessionCoordinator) {
        try {
            val response = profileApiService.getMyProfileSignals()
            return@executeAuthorizedApiCall ApiResult.Success(parseProfileSignalsSnapshot(response))
        } catch (error: HttpException) {
            if (error.code() != 404) {
                throw error
            }
        }

        loadLegacyWorkerProfileSnapshot(workerId)
    }

    private suspend fun loadLegacyWorkerProfileSnapshot(workerId: String): ApiResult<WorkerProfileSnapshot> =
        coroutineScope {
            val credentialsDeferred = async { fetchOptional { profileApiService.getMyCredentials() } }
            val availabilityDeferred = async { fetchOptional { profileApiService.getWorkerAvailability(workerId) } }
            val completenessDeferred = async { fetchOptional { profileApiService.getWorkerCompleteness(workerId) } }
            val portfolioDeferred = async {
                fetchOptional {
                    profileApiService.getWorkerPortfolio(
                        workerId = workerId,
                        query = mapOf("limit" to "6"),
                    )
                }
            }

            val profileResponse = profileApiService.getProfile()
            val credentialsResponse = credentialsDeferred.await()
            val availabilityResponse = availabilityDeferred.await()
            val completenessResponse = completenessDeferred.await()
            val portfolioResponse = portfolioDeferred.await()
            ApiResult.Success(
                WorkerProfileSnapshot(
                    profile = parseProfile(profileResponse),
                    credentials = parseCredentials(credentialsResponse),
                    availability = parseAvailability(availabilityResponse),
                    completeness = parseCompleteness(completenessResponse),
                    portfolio = parsePortfolio(portfolioResponse),
                    partialWarnings = buildList {
                        if (credentialsResponse == null) add("Your certificates could not load. Job matches may be less accurate.")
                        if (availabilityResponse == null) add("Your work time could not load. Job matches may be less up to date.")
                        if (completenessResponse == null) add("Your profile check could not load.")
                        if (portfolioResponse == null) add("Your past work could not load. Hirers may see less proof.")
                    },
                ),
            )
        }

    private fun parseProfileSignalsSnapshot(response: JsonObject): WorkerProfileSnapshot {
        val raw = response.nestedObject("data") ?: response
        return WorkerProfileSnapshot(
            profile = parseProfileObject(raw.nestedObject("profile") ?: JsonObject(emptyMap())),
            credentials = parseCredentials(raw.nestedObject("credentials")),
            availability = parseAvailability(raw.nestedObject("availability")),
            completeness = parseCompleteness(raw.nestedObject("completeness")),
            portfolio = parsePortfolio(raw.nestedObject("portfolio")),
        )
    }

    private suspend fun fetchOptional(block: suspend () -> JsonObject): JsonObject? =
        runCatching { block() }.getOrNull()

    private fun parseProfile(response: JsonObject): WorkerRecommendationProfile {
        val raw = response.nestedObject("data") ?: response
        return parseProfileObject(raw)
    }

    private fun parseProfileObject(raw: JsonObject): WorkerRecommendationProfile {
        return WorkerRecommendationProfile(
            id = raw.string("id") ?: raw.string("_id"),
            firstName = raw.string("firstName") ?: "",
            lastName = raw.string("lastName") ?: "",
            email = raw.string("email") ?: "",
            phone = raw.string("phone") ?: "",
            bio = raw.string("bio") ?: "",
            location = raw.string("location")
                ?: listOfNotNull(raw.string("city"), raw.string("state"), raw.string("country"))
                    .filter { it.isNotBlank() }
                    .joinToString(", "),
            profession = raw.string("profession") ?: "",
            hourlyRate = raw.double("hourlyRate"),
            currency = raw.string("currency") ?: "GHS",
            experienceLevel = raw.string("experienceLevel"),
            yearsOfExperience = raw.int("yearsOfExperience"),
            skills = raw.nestedArray("skills").toStringList(),
            isEmailVerified = raw.bool("isEmailVerified") ?: false,
            isPhoneVerified = raw.bool("isPhoneVerified") ?: false,
        )
    }

    private fun parseCredentials(response: JsonObject?): WorkerCredentials {
        val raw = response?.nestedObject("data") ?: response ?: return WorkerCredentials()
        return WorkerCredentials(
            skills = raw.nestedArray("skills")?.mapIndexedNotNull { index, value ->
                val obj = value as? JsonObject
                val name = when {
                    obj != null -> obj.string("name") ?: obj.string("label")
                    value is JsonPrimitive -> value.content.takeIf { it.isNotBlank() }
                    else -> null
                } ?: return@mapIndexedNotNull null
                CredentialSkill(
                    id = obj?.string("id") ?: obj?.string("_id") ?: "skill-$index-$name",
                    name = name,
                    category = obj?.string("category") ?: "general",
                    proficiencyLevel = obj?.string("proficiencyLevel") ?: obj?.string("level") ?: "intermediate",
                    yearsOfExperience = obj?.int("yearsOfExperience") ?: 0,
                    isVerified = obj?.bool("isVerified") ?: false,
                )
            } ?: emptyList(),
            licenses = parseCredentialItems(raw.nestedArray("licenses")),
            certifications = parseCredentialItems(raw.nestedArray("certifications")),
        )
    }

    private fun parseCredentialItems(items: JsonArray?): List<CredentialItem> =
        items?.mapIndexedNotNull { index, value ->
            val obj = value as? JsonObject ?: return@mapIndexedNotNull null
            val name = obj.string("name") ?: return@mapIndexedNotNull null
            CredentialItem(
                id = obj.string("id") ?: obj.string("_id") ?: "credential-$index-$name",
                name = name,
                issuingOrganization = obj.string("issuingOrganization") ?: obj.string("issuer") ?: "",
                issueDate = obj.string("issueDate") ?: obj.string("issuedAt"),
                expiryDate = obj.string("expiryDate") ?: obj.string("expiresAt"),
                status = obj.string("status") ?: if (obj.bool("isVerified") == true) "verified" else "pending",
                isVerified = obj.bool("isVerified") ?: false,
            )
        } ?: emptyList()

    private fun parseAvailability(response: JsonObject?): WorkerAvailability {
        val raw = response?.nestedObject("data") ?: response ?: return WorkerAvailability()
        val schedule = (raw.nestedArray("schedule") ?: raw.nestedArray("daySlots"))
            ?.mapNotNull dayEntry@{ value ->
                val obj = value as? JsonObject ?: return@dayEntry null
                val day = obj.string("day") ?: return@dayEntry null
                AvailabilityDay(
                    day = day,
                    available = obj.bool("available") ?: obj.nestedArray("slots")?.isNotEmpty() == true,
                    slots = obj.nestedArray("slots")?.mapNotNull slotEntry@{ slotValue ->
                        val slot = slotValue as? JsonObject ?: return@slotEntry null
                        val start = slot.string("start") ?: return@slotEntry null
                        val end = slot.string("end") ?: return@slotEntry null
                        AvailabilitySlot(start = start, end = end)
                    } ?: emptyList(),
                )
            }
            ?: emptyList()

        return WorkerAvailability(
            status = raw.string("status") ?: "not_set",
            isAvailable = raw.bool("isAvailable") ?: (raw.string("status") == "available"),
            timezone = raw.string("timezone") ?: "Africa/Accra",
            schedule = schedule,
            nextAvailable = raw.string("nextAvailable"),
            lastUpdated = raw.string("lastUpdated"),
            message = raw.string("message"),
        )
    }

    private fun parseCompleteness(response: JsonObject?): WorkerCompleteness {
        val raw = response?.nestedObject("data") ?: response ?: return WorkerCompleteness()
        return WorkerCompleteness(
            completionPercentage = raw.int("completionPercentage") ?: raw.int("percentage") ?: 0,
            requiredCompletion = raw.int("requiredCompletion") ?: 0,
            optionalCompletion = raw.int("optionalCompletion") ?: 0,
            missingRequired = raw.nestedArray("missingRequired").toStringList(),
            missingOptional = raw.nestedArray("missingOptional").toStringList(),
            recommendations = raw.nestedArray("recommendations").toStringList(),
        )
    }

    private fun parsePortfolio(response: JsonObject?): WorkerPortfolio {
        val raw = response?.nestedObject("data") ?: response ?: return WorkerPortfolio()
        val items = raw.nestedArray("portfolioItems")?.mapNotNull { value ->
            val obj = value as? JsonObject ?: return@mapNotNull null
            val id = obj.string("id") ?: obj.string("_id") ?: return@mapNotNull null
            PortfolioProject(
                id = id,
                title = obj.string("title") ?: "Untitled project",
                description = obj.string("description") ?: "",
                projectType = obj.string("projectType") ?: "professional",
                skillsUsed = obj.nestedArray("skillsUsed").toStringList(),
                location = obj.string("location"),
                clientRating = obj.double("clientRating"),
                status = obj.string("status") ?: "draft",
                isFeatured = obj.bool("isFeatured") ?: false,
                createdAt = obj.string("createdAt"),
            )
        } ?: emptyList()
        val stats = raw.nestedObject("stats")
        return WorkerPortfolio(
            items = items,
            totalCount = stats?.int("total") ?: items.size,
            publishedCount = stats?.int("published") ?: items.count { it.status.equals("published", ignoreCase = true) },
        )
    }
}

private fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)
    ?.takeIf { it != JsonNull }
    ?.content
    ?.takeIf { it.isNotBlank() }

private fun JsonObject.int(key: String): Int? = (this[key] as? JsonPrimitive)?.content?.toIntOrNull()

private fun JsonObject.double(key: String): Double? = (this[key] as? JsonPrimitive)?.content?.toDoubleOrNull()

private fun JsonObject.bool(key: String): Boolean? = (this[key] as? JsonPrimitive)?.content?.toBooleanStrictOrNull()

private fun JsonObject.nestedObject(key: String): JsonObject? = this[key] as? JsonObject

private fun JsonObject.nestedArray(key: String): JsonArray? = this[key] as? JsonArray

private fun JsonArray?.toStringList(): List<String> =
    this?.mapNotNull { value ->
        when (value) {
            is JsonPrimitive -> value.content.takeIf { it.isNotBlank() }
            is JsonObject -> value.string("name") ?: value.string("label") ?: value.string("type")
            JsonNull -> null
            else -> null
        }
    }?.filter { it.isNotBlank() }?.distinct() ?: emptyList()